while-loops/problem-set-while-loops.md

```diff
@@ -0,0 +1,849 @@
+
+The `while` loop iterates over the contents of the list of famous cats in order, printing each out, one by one. The correct `for` loop does the same thing, but notice how much clearer it is. We don't have to set up the counter. We don't have to compare it to the length of the list. We don't have to index into the list values. And we don't have to advance the counter. `while` loops are very flexible, but for some tasks, like iterating over lists and dictionaries in order, a `for` loop is usually more appropriate.
+
+<br />
+
+Compare the correct `for` loop to the other answers and try explaining why they don't work. Try running them in the Python interpreter and notice the errors it reports for each.
+
+##### !end-explanation
+
+### !end-challenge
+<!-- ======================= END CHALLENGE ======================= -->
+<!-- prettier-ignore-end -->
+
+<!-- prettier-ignore-start -->
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+
+### !challenge
+
+* type: ordering
+* id: 879e5790-de4c-4a51-ac38-0330ffb98a67
+* title: While Loops
+
+##### !question
+
+Reorder the lines of code below to produce a `while` loop with the following behavior:
+
+- Prompt the user for input.
+- Ignore the input if the user enters 'skip.'
+- End the loop if the user enters 'done.'
+- Otherwise, print a message telling the user what they input.
+
+Assume that each line would be indented correctly.
+
+##### !end-question
+
+##### !answer
+
+1. while True:
+1.     user_input = input('enter input: ')
+1.     if user_input == 'done':
+1.         break
+1.     elif user_input == 'skip':
+1.         continue
+1.     print(f'you entered {user_input}')
+
+##### !end-answer
+
+##### !explanation
+
+This implementation uses `while True: ...` to run the loop forever. It exits the loop by using the `break` keyword. And it skips the `print` by using the `continue` keyword. There are many other approaches that are just as valid. For an extra challenge, try implementing this loop yourself without using `break` or `continue`!
+
+##### !end-explanation
+
+### !end-challenge
+
+<!-- ======================= END CHALLENGE ======================= -->
+<!-- prettier-ignore-end -->
+
+
+<!-- prettier-ignore-start -->
+<!-- >>>>>>>>>>>>>>>>>>>>>> BEGIN CHALLENGE >>>>>>>>>>>>>>>>>>>>>> -->
+<!-- Replace everything in square brackets [] and remove brackets  -->
+
+### !challenge
+
+* type: code-snippet
+* language: python3.6
+* id: 85dffbd3-2030-44b4-8b9f-cc86f0c7e9d1
+* title: While Loops
+
+##### !question
+
+Let's implement a function called `sidewinder`. It will take a parameter `number_list` which contains a list of the _next_ index in the list to visit. Starting from position `0`, we will look in the current position, get the value stored there, and use it as the _next_ position to visit. We will continue doing this until we find a value of `0`, which means we have found the end of the index chain. Then we'll return the position where we found the `0`.
```

Comment:
I'm making a suggestion for how to re-word this function. Summary of recommendations:

- Lead the description with what the function is responsible for doing ultimately, even before details of how to get there
- Makes up verbiage that says this is a fictional algorithm, to help reassure that this is totally a made up algorithm. I didn't really do my research to find out how true that is, but I would want to assure students that this is something made up for an exercise, not an algorithm that they should memorize.
- `number_list` _is_ a list of those indices (not contains)
- Words "starting from position `0`" as a rule necessary to solving the problem, rather implies it
- Formats it as a bulleted list to help students turn it into code
- I started to use the word "index" so I added a parenthetical (index) to the final use of "position"

```suggestion
Let's implement a function called `sidewinder`. This function is responsible for finding the _index_ in a list that has the value `0` when following the imaginary "sidewinder algorithm."

In order to find the correct index, the function:
  - Takes a parameter `number_list` which is a list of "next" indices to visit
  - The function always starts at index `0`
  - The function will look in the current position, get the value stored there. Then, it will use that value as the _next_ position to visit
  
The function continues doing this until we find a value of `0`, which means we have found the end of the index chain. Then we'll return the _position_ (index) where we found the `0`.
```

Reply:
Love it! Just changed "when following" to "by following"

```suggestion
Let's implement a function called `sidewinder`. This function is responsible for finding the _index_ in a list that has the value `0` by following the imaginary "sidewinder algorithm."

In order to find the correct index, the function:
  - Takes a parameter `number_list` which is a list of "next" indices to visit
  - The function always starts at index `0`
  - The function will look in the current position, get the value stored there. Then, it will use that value as the _next_ position to visit
  
The function continues doing this until we find a value of `0`, which means we have found the end of the index chain. Then we'll return the _position_ (index) where we found the `0`.
```